src/box/lua/schema.lua

```diff
@@ -3453,35 +3507,58 @@ local function grant(uid, name, privilege, object_type,
-    local tuple = _vpriv:get{uid, object_type, oid}
-    local old_privilege
-    if tuple ~= nil then
-        old_privilege = tuple.privilege
-    else
-        old_privilege = 0
-    end
-    privilege_hex = bit.bor(privilege_hex, old_privilege)
-    privilege_hex = tonumber(ffi.cast('uint32_t', privilege_hex))
+    local origins = origins_from_tuple(_vpriv:get({uid, object_type, oid}))
+
     -- do not execute a replace if it does not change anything
     -- XXX bug if we decide to add a grant option: new grantor
     -- replaces the old one, old grantor is lost
-    if privilege_hex ~= old_privilege then
-        _priv:replace{options.grantor, uid, object_type, oid, privilege_hex}
-    elseif not options.if_not_exists then
-            if object_type == 'role' and object_name ~= '' and
-               privilege == 'execute' then
-                box.error(box.error.ROLE_GRANTED, name, object_name)
-            else
-                if not is_singleton_object_type(object_type) then
-                    object_name = string.format(" '%s'", object_name)
-                end
-                box.error(box.error.PRIV_GRANTED, name, privilege,
-                          object_type, object_name)
-            end
+    local old_privilege = origins[options._origin] or 0
+    if bit.band(bit.bnot(old_privilege), privilege_hex) ~= 0 then
+        -- Update given privileges by origin.
+        origins[options._origin] = bit.bor(old_privilege, privilege_hex)
+    else
+        -- No new privileges can be given.
+        if options.if_not_exists then
+            return
+        end
+        grant_error(name, object_name, object_type, privilege)
+    end
+
+    local new_privilege = privilege_from_origins(origins)
+    local opts = opts_from_origins(origins)
+    _priv:replace({options.grantor, uid, object_type, oid, new_privilege, opts})
+end
+
+local function revoke_error(name, object_name, object_type, privilege, origin,
+                            tuple)
+    local total_privileges = tuple == nil and 0 or tuple.privilege
+    local origin_name = origin == '' and 'default' or origin
+    local grantee_type = box.space._user.index.name:get(name).type == 'user' and
+                         'User' or 'Role'
+    local new_privilege = privilege_check(privilege, object_type)
+    if object_type == 'role' and object_name ~= '' and
+       privilege == 'execute' then
+        if bit.band(new_privilege, total_privileges) == 0 then
+            box.error(box.error.ROLE_NOT_GRANTED, name, object_name)
+        end
+        box.error(box.error.ROLE_NOT_GRANTED_BY_ORIGIN, grantee_type, name,
+                  object_name, origin_name)
+    end
+    if bit.band(new_privilege, total_privileges) == 0 then
+        box.error(box.error.PRIV_NOT_GRANTED, name, privilege, object_type,
+                  object_name)
     end
+    box.error(box.error.PRIV_NOT_GRANTED_BY_ORIGIN, grantee_type, name,
+              privilege, object_type, object_name, origin_name)
```

Comment:
It may be better to add a new payload parameter to the existing errors `PRIV_NOT_GRANTED` and `ROLE_NOT_GRANTED` instead of introducing the new error codes.